File: apps/api/app/schemas/forecast.py
"""Pydantic schemas for the forecast API (T8), matching the issue's API
contract shape 1:1 (field names/nesting) with two documented deviations:

- `recurring_id` is a `str` (the RecurringPayment's canonical merchant
  name, see `app/models/recurring_payment.py`), not an `int` DB PK -
  consistent with the "no database" decision throughout this feature.
- `Assumptions` has an extra `notes: list[str]` field beyond the issue's
  example JSON, used for the free-text edge-case hints the issue asks
  for ("Lohntag nicht erkennbar", "weniger als 6 Monate Historie", ...)
  without needing a dedicated boolean per edge case.

Built together with T7 (`app/services/forecast_service.py`), which
returns these directly - there's no separate internal representation to
keep in sync with the API contract.
"""

from __future__ import annotations

from datetime import date
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

from app.models.recurring_payment import Interval

Horizon = Literal["next_salary", "30d", "90d", "365d"]


class NextSalary(BaseModel):
    date: date
    amount_chf: float


class FreeToSpend(BaseModel):
    expected_chf: float
    lower_chf: float
    upper_chf: float


class TightDate(BaseModel):
    date: date
    days_until: int
    days_before_salary: int | None
    projected_balance_chf: float


class KnownPayment(BaseModel):
    date: date
    label: str
    amount_chf: float
    """Positive magnitude regardless of expense/income, matching
    `RecurringPayment.amount_chf`'s convention."""
    category: str | None
    recurring_id: str | None
    """None for one-off adjustments (`simulate`, `type: "one_off"|"add_recurring"`
    before they've been projected) that don't map back to a detected
    RecurringPayment."""


class SeriesPoint(BaseModel):
    date: date
    expected_chf: float
    lower_chf: float
    upper_chf: float


class Assumptions(BaseModel):
    variable_baseline_method: str
    band_method: str
    excluded_outliers: list[str]
    interest_applied: bool = False
    salary_day_detected: bool
    variable_baseline_months_used: int
    notes: list[str] = Field(default_factory=list)


class ForecastResponse(BaseModel):
    as_of: date
    horizon: Horizon
    horizon_end: date
    opening_balance_chf: float
    next_salary: NextSalary | None
    free_to_spend: FreeToSpend
    tight_date: TightDate | None
    known_payments: list[KnownPayment]
    series: list[SeriesPoint]
    assumptions: Assumptions


# --- simulate() request -----------------------------------------------


class CancelRecurring(BaseModel):
    type: Literal["cancel_recurring"] = "cancel_recurring"
    recurring_id: str
    effective_from: date | None = None
    """Default: cancelled immediately. If set, occurrences before this
    date still happen - issue's "Gekündigtes Abo, das noch bis
    Periodenende läuft" edge case."""


class AdjustRecurring(BaseModel):
    type: Literal["adjust_recurring"] = "adjust_recurring"
    recurring_id: str
    delta_chf: float
    """Added to the current magnitude, e.g. +200 for "Miete +200"."""


class AddRecurring(BaseModel):
    type: Literal["add_recurring"] = "add_recurring"
    label: str
    amount_chf: float
    interval: Interval
    start_date: date


class OneOff(BaseModel):
    type: Literal["one_off"] = "one_off"
    label: str
    amount_chf: float
    date: date


Adjustment = Annotated[
    Union[CancelRecurring, AdjustRecurring, AddRecurring, OneOff],
    Field(discriminator="type"),
]


class SimulateRequest(BaseModel):
    horizon: Horizon = "next_salary"
    as_of: date | None = None
    adjustments: list[Adjustment]


class DiffPoint(BaseModel):
    date: date
    diff_chf: float


class Diff(BaseModel):
    monthly_chf: float
    cumulative_series: list[DiffPoint]
    total_at_horizon_chf: float
    tight_date_shift_days: int | None
    """Positive = scenario's tight_date is later (more buffer). `None` if
    either baseline or scenario never hits the buffer within the horizon."""


class SimulateResponse(BaseModel):
    baseline: ForecastResponse
    scenario: ForecastResponse
    diff: Diff

File: apps/web/src/app/core/forecast.ts
import { httpResource } from '@angular/common/http';
import { Service, computed, signal } from '@angular/core';

export type Horizon = 'next_salary' | '30d' | '90d' | '365d';
export type Interval = 'monthly' | 'quarterly' | 'yearly' | 'irregular';

export interface SeriesPoint {
  date: string;
  expected_chf: number;
  lower_chf: number;
  upper_chf: number;
}

export interface KnownPayment {
  date: string;
  label: string;
  amount_chf: number;
  category: string | null;
  recurring_id: string | null;
}

export interface ForecastResult {
  as_of: string;
  horizon: Horizon;
  horizon_end: string;
  opening_balance_chf: number;
  next_salary: { date: string; amount_chf: number } | null;
  free_to_spend: { expected_chf: number; lower_chf: number; upper_chf: number };
  tight_date: {
    date: string;
    days_until: number;
    days_before_salary: number | null;
    projected_balance_chf: number;
  } | null;
  known_payments: KnownPayment[];
  series: SeriesPoint[];
  assumptions: {
    variable_baseline_method: string;
    band_method: string;
    excluded_outliers: string[];
    interest_applied: boolean;
    salary_day_detected: boolean;
    variable_baseline_months_used: number;
    notes: string[];
  };
}

export interface Diff {
  monthly_chf: number;
  cumulative_series: { date: string; diff_chf: number }[];
  total_at_horizon_chf: number;
  tight_date_shift_days: number | null;
}

export interface RecurringPayment {
  recurring_id: string;
  merchant: string;
  category_main: string | null;
  amount_chf: number;
  interval: Interval;
  flow: 'expense' | 'income';
  is_active: boolean;
}

export type AdjustmentPayload =
  | { type: 'cancel_recurring'; recurring_id: string }
  | { type: 'adjust_recurring'; recurring_id: string; delta_chf: number }
  | { type: 'add_recurring'; label: string; amount_chf: number; interval: Interval; start_date: string }
  | { type: 'one_off'; label: string; amount_chf: number; date: string };

/** An adjustment plus the chip label the UI shows for it. */
export interface Adjustment {
  id: string;
  label: string;
  payload: AdjustmentPayload;
}

export const HORIZONS: readonly { value: Horizon; label: string }[] = [
  { value: 'next_salary', label: 'Bis Lohn' },
  { value: '30d', label: '30 Tage' },
  { value: '90d', label: '90 Tage' },
  { value: '365d', label: '1 Jahr' },
];

type SimulateResult = { baseline: ForecastResult; scenario: ForecastResult; diff: Diff };

const ODATA = '/odata';

/**
 * Owns the horizon, the active scenario adjustments, and the forecast
 * they produce. One resource serves both cases: with no adjustments it
 * reads the baseline (`GetForecast`), otherwise it asks for baseline and
 * scenario in one round trip (`Simulate`) so both curves can be drawn
 * without a second request.
 */
@Service()
export class Forecast {
  readonly horizon = signal<Horizon>('next_salary');
  readonly adjustments = signal<Adjustment[]>([]);

  private readonly data = httpResource<ForecastResult | SimulateResult>(() => {
    const horizon = this.horizon();
    const adjustments = this.adjustments();
    return adjustments.length
      ? {
          url: `${ODATA}/Simulate`,
          method: 'POST',
          body: { horizon, adjustments: adjustments.map((adjustment) => adjustment.payload) },
        }
      : { url: `${ODATA}/GetForecast`, method: 'GET', params: { horizon } };
  });

  /** Active expense subscriptions, for the cancel picker. */
  readonly recurring = httpResource<{ value: RecurringPayment[] }>(() => ({
    url: `${ODATA}/RecurringPayments`,
    params: { $filter: "is_active eq true and flow eq 'expense'", $orderby: 'amount_chf desc' },
  }));

  readonly isLoading = this.data.isLoading;
  readonly error = this.data.error;

  private readonly result = computed(() => (this.data.hasValue() ? this.data.value() : null));

  readonly baseline = computed<ForecastResult | null>(() => {
    const result = this.result();
    if (!result) return null;
    return 'baseline' in result ? result.baseline : result;
  });

  /** Non-null only while a scenario is active. */
  readonly scenario = computed<ForecastResult | null>(() => {
    const result = this.result();
    return result && 'scenario' in result ? result.scenario : null;
  });

  readonly diff = computed<Diff | null>(() => {
    const result = this.result();
    return result && 'diff' in result ? result.diff : null;
  });

  /** The curve the headline numbers describe: scenario if there is one. */
  readonly current = computed(() => this.scenario() ?? this.baseline());

  readonly expenseSubscriptions = computed(() =>
    this.recurring.hasValue() ? this.recurring.value().value : [],
  );

  /** The largest recurring expenses, used to build data-driven presets. */
  readonly topMonthly = computed(() =>
    this.expenseSubscriptions().filter((payment) => payment.interval === 'monthly'),
  );

  add(adjustment: Adjustment): void {
    this.adjustments.update((current) =>
      current.some((existing) => existing.id === adjustment.id) ? current : [...current, adjustment],
    );
  }

  remove(id: string): void {
    this.adjustments.update((current) => current.filter((adjustment) => adjustment.id !== id));
  }

  reset(): void {
    this.adjustments.set([]);
  }
}
